### PR TITLE
Fast membrane current calculation

### PR DESCRIPTION
**Pull request by iomaganaris**
Related to https://github.com/BlueBrain/mod2c/pull/32 and https://github.com/neuronsimulator/nrn/pull/272

### CODE REVIEW

File: coreneuron/nrnoc/fast_imem.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            free(nrn_threads[i].nrn_fast_imem);
+            nrn_threads[i].nrn_fast_imem = nullptr;
+        }
+    }
+}
+
+void nrn_fast_imem_alloc() {
+    if (nrn_use_fast_imem) {
+        fast_imem_free();
+        for (int i=0; i < nrn_nthread; ++i) {
```
[the remark this answers is not in the document]
Comment (iomaganaris) [reply coreneuron/nrnoc/fast_imem.cpp]: True.. I got carried away from Neuron.

File: coreneuron/nrniv/nrn_setup.cpp [hunk 2]
```diff
@@ -913,16 +919,24 @@ int nrn_i_layout(int icnt, int cnt, int isz, int sz, int layout) {
...
+    if (mtype == voltage) {  // voltage
         int v0 = nt._actual_v - nt._data;
         int ix = index;  // relative to _actual_v
         nrn_assert((ix >= 0) && (ix < nt.end));
         if (nt._permute) {
             node_permute(&ix, 1, nt._permute);
         }
         return nt._data + (v0 + ix);                // relative to nt._data
+    } else if (mtype == i_membrane_) {  // membrane current from fast_imem calculation
+            int i_mem = nt.nrn_fast_imem->nrn_sav_rhs - nt._data;
```
[the remark this answers is not in the document]
Comment (iomaganaris) [reply coreneuron/nrniv/nrn_setup.cpp]: Haha good catch. Something was not pretty there :+1:

**Comment by nrnhines** [comment]
retest

**Comment by nrnhines** [comment]
Please retest

File: coreneuron/nrnoc/fast_imem.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+#include "coreneuron/nrnoc/fast_imem.h"
+#include "coreneuron/nrniv/memory.h"
+#include "coreneuron/nrnmpi/nrnmpi.h"
+
+namespace coreneuron {
+
+extern int nrn_nthread;
+extern NrnThread *nrn_threads;
+bool nrn_use_fast_imem;
+static int fast_imem_nthread_ = 0;
```
Comment (pramodk) [inline coreneuron/nrnoc/fast_imem.cpp]: not used?

File: coreneuron/nrnoc/fast_imem.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+static int *fast_imem_size_ = nullptr;
+static NrnFastImem* fast_imem_;
```
Comment (pramodk) [inline coreneuron/nrnoc/fast_imem.cpp]: both unused?

File: coreneuron/nrnoc/fast_imem.cpp (hunk 3 above)
Comment (iomaganaris) [reply coreneuron/nrnoc/fast_imem.cpp]: True about all of them.. I forgot to delete them :+1:

File: coreneuron/nrnoc/multicore.h [hunk 5]
```diff
@@ -59,6 +59,11 @@ struct NrnThreadBAList {
     NrnThreadBAList* next;
 };
 
+typedef struct NrnFastImem {
+    double* nrn_sav_rhs;
+    double* nrn_sav_d;
+} NrnFastImem;
```
[the remark this answers is not in the document]
Comment (iomaganaris) [reply coreneuron/nrnoc/multicore.h]: Ah true. I got carried away from C again..

**Pull request merged**